subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_client.c

```diff
@@ -0,0 +1,542 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define LOG_MODULE_NAME mcumgr_client_img_grp
+#define LOG_LEVEL	CONFIG_LWM2M_LOG_LEVEL
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <zcbor_common.h>
+#include <zcbor_decode.h>
+#include <zcbor_encode.h>
+#include <pm_config.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/net/buf.h>
+#include <zephyr/mgmt/mcumgr/mgmt/mgmt.h>
+#include <zephyr/mgmt/mcumgr/smp/smp.h>
+#include <zephyr/mgmt/mcumgr/smp/smp_client.h>
+#include <zephyr/mgmt/mcumgr/transport/smp.h>
+#include <zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt.h>
+#include <zephyr/mgmt/mcumgr/grp/img_mgmt/img_mgmt_client.h>
+
+#include <mgmt/mcumgr/util/zcbor_bulk.h>
+#include <mgmt/mcumgr/transport/smp_internal.h>
+#include <mgmt/mcumgr/grp/img_mgmt/img_mgmt_priv.h>
+
+/* Mimimal number of ZCBOR encoder states to provide full encoder functionality plus one for backup
+ */
+#define CBOR_ENCODER_STATE_NUM 3
+
+#define MCUMGR_IMAGE_LIST_SIZE 2
+
+struct upload_progress {
+	size_t image_size;
+	size_t offset;
+	size_t request_length;
+	size_t writed_length;
+	int image_num;
+	int upload_status;
+};
+
+static struct upload_progress upload_state;
+static mgmt_client_gr_image_fn client_gr_image_fn;
+static struct mcumgr_image_data image_info[MCUMGR_IMAGE_LIST_SIZE];
```

Comment:
This is not multiple transport safe (does not need changing for first implementation, just a note)

Reply:
I will replace this in callback remove update state. User give buffer in future

Reply:
Static buffer removed user gives buffer pointer for storage image list data.